### ISSUE TITLE
IMPORTS-BUT-NO-CALLS: Python PyPI-to-import name mapping edge cases after #216 fix

### ISSUE THREAD

**Issue by kotakanbe** [future-architect/uzomuzo-oss#246]
## Anomaly

**Type**: IMPORTS-BUT-NO-CALLS / FALSE-UNUSED
**Language**: Python (PyPI)
**SBOM tool**: trivy v0.69.3
**Date**: 2026-04-08
**Related**: Follow-up to #216 (closed)

## Context

Issue #216 fixed the primary Python package-name vs import-name mismatch for `python-multipart` → `multipart`. However, the same root cause affects other well-known PyPI packages where the package name differs from the import module name.

### 1. `pyyaml` → `yaml` (ansible/ansible)

```python
# Ansible imports yaml extensively (~15+ files)
import yaml
yaml.safe_load(data)
yaml.dump(obj)
```

- **PyPI package**: `pyyaml`
- **Python import**: `yaml`
- **Diet result**: `import_file_count: 0, call_site_count: 0` — classified as trivial/unused
- **Reality**: One of Ansible's most heavily used dependencies

### 2. `click-didyoumean` → `click_didyoumean` (celery/celery)

```python
# celery/bin/celery.py
from click_didyoumean import DYMGroup
```

- import is detected (1 file), but `DYMGroup` is used as `cls=DYMGroup` keyword argument — class reference, not a function call
- This is both a name mismatch issue AND a call site detection pattern gap (class-as-argument)

## Known PyPI → Import Name Mismatches

A lookup table for well-known cases would cover the most impactful false negatives:

| PyPI name | Import name | Prevalence |
|-----------|------------|------------|
| `pyyaml` | `yaml` | Very high |
| `pillow` | `PIL` | Very high |
| `scikit-learn` | `sklearn` | Very high |
| `beautifulsoup4` | `bs4` | High |
| `python-dateutil` | `dateutil` | High |
| `python-multipart` | `multipart` | Medium (fixed in #216) |
| `attrs` | `attr` | Medium |
| `protobuf` | `google.protobuf` | Medium |
| `opencv-python` | `cv2` | High |
| `python-dotenv` | `dotenv` | High |

## Suggested Fix

Add a static mapping table in the coupling analyzer for well-known PyPI→import name mismatches. This covers the 80/20 case. A more comprehensive approach would be to use the PyPI metadata `top_level.txt` field, but the static table handles the most common cases with minimal complexity.

## Reproduction

```bash
git clone --depth 1 https://github.com/ansible/ansible.git /tmp/diet-trial-ansible
# Generate SBOM (Trivy has trouble with ansible's setup — manual SBOM may be needed)
uzomuzo-diet --sbom /tmp/sbom.json --source /tmp/diet-trial-ansible --format json
# Check entry for pyyaml — will show 0 imports, 0 calls
```

## Evidence

Full diet JSON at `docs/case-studies/uzomuzo-diet/python/ansible-trivy-2026-04-08.json`.

---
*Filed by `/diet-trial` batch run (2026-04-08)*

**Comment by kotakanbe** [comment]
## Additional evidence from diet-fuzz run 2026-04-08T2354

**Affected projects**: 4/5 (gptme, inference, advanced-alchemy, dagster)
**SBOM tools**: trivy v0.69.3, syft v1.42.4, cdxgen v12.1.4

### Key IBNC patterns found

| Project | Dependency | Tool | Files | Pattern |
|---------|-----------|------|-------|---------|
| dagster | remark-gfm, rehype-highlight, etc. (JS deps) | trivy | 1-4 | JS deps in Python monorepo — trivy includes JS workspace deps |
| dagster | project-components-pdf-extraction | syft | 2 | Internal/example package reference |
| gptme | click-default-group | trivy, syft, cdxgen | 1 | Python CLI extension — loaded via click group plugin mechanism |
| gptme | typing-extensions | trivy, cdxgen | 8 | Used via `TYPE_CHECKING` conditional imports |
| gptme | @icons-pack/react-simple-icons | trivy, cdxgen | 1 | JS dep in Python project with web frontend |
| advanced-alchemy | pytest-databases, cattrs, pytest-mock | syft | 1-6 | Test dependencies — used via pytest plugins, not direct calls |
| inference | @tailwindcss/line-clamp | trivy | 1 | CSS/JS dep in Python project |

### Observations

1. **Cross-language dep contamination**: dagster and gptme are Python monorepos with JS frontends. SBOM tools (especially trivy) include JS workspace deps, which the Python coupling analyzer cannot trace.
2. **Pytest plugin pattern**: `pytest-databases`, `pytest-mock` are invoked by pytest's plugin system, not direct imports with call sites.
3. **`typing-extensions` pattern**: Imported inside `if TYPE_CHECKING:` blocks — the tree-sitter analyzer sees the import but the call sites use the conditional import differently.
4. **`click-default-group`**: Registered as a click group class, call site is the decorator/class reference — not a standard function call.

### Summary

| Metric | Value |
|--------|-------|
| Total Python IBNC (real) | 30 |
| Known limitation (config-driven) | 9 |
| Likely parser improvements needed | ~5 (click plugin, typing-extensions, pytest plugin) |

---
*Added by `/diet-fuzz` run 2026-04-08T2354*

**Comment by kotakanbe** [comment]
## Additional evidence from diet-fuzz round 2 (2026-04-09T0115)

**Affected projects**: 2/3 successful (flask, fastapi — requests had only 1 dep)
**SBOM tools**: trivy v0.69.3, syft v1.42.4, cdxgen v12.1.4
**Total IBNC**: 3 real anomalies

### Findings

| Project | Dep | Tool | Files | Score | Pattern |
|---------|-----|------|-------|-------|---------|
| flask | cryptography | syft | 1 | 0.096 | Conditional import in try/except — already filed as #243 |
| fastapi | email-validator | trivy | 1 | 0.025 | Optional dependency — `try: import email_validator` |
| fastapi | python-multipart | trivy | 1 | 0.019 | Optional dependency — imported in type-checking or conditional block |

### Pattern confirmation

All 3 Python IBNC in round 2 are **optional/conditional imports**:
- `cryptography` in flask: `try: import cryptography` for optional HTTPS support
- `email-validator` in fastapi: optional validation dependency
- `python-multipart` in fastapi: optional form parsing dependency

This confirms the pattern from #243 — Python's `try/except ImportError` pattern for optional deps is a systematic IBNC source. The tree-sitter analyzer detects the `import` inside the `try` block but cannot determine that the call sites are inside the corresponding `except`-guarded code paths.

### Skipped projects (2/5)

- **httpie/cli**: No lockfile (only `pyproject.toml`) — all tools produced empty SBOMs
- **encode/django-rest-framework**: Same issue — no lockfile

This confirms the round 1 finding about Python projects needing lockfiles for SBOM generation.

---
*Added by `/diet-fuzz` round 2 run 2026-04-09T0115*

**Comment by kotakanbe** [comment]
## Massive evidence from diet-fuzz round 4 (2026-04-11T1310)

**Project**: home-assistant/core (trivy v0.69.3)
**Impact**: **83 confirmed name mismatch cases, 1,003 missed import files**

Of 146 deps classified as "unused" in home-assistant, **83 are actually imported in source code** but diet misses them because the PyPI package name differs from the Python import name.

### Mismatch categories

| Category | Count | Examples |
|----------|-------|---------|
| py-prefix removal | 24 | `pyserial`→`serial` (38 imports), `pyswitchbot`→`switchbot` (76) |
| Completely different names | 41 | `pillow`→`PIL` (20), `matter-python-client`→`matter_server` (93), `hap-python`→`pyhap` (55) |
| Google/Azure namespaces | 9 | `google-genai`→`google.genai` (19), `azure-eventhub`→`azure.eventhub` (8) |
| Suffix -py/-python removal | 8 | `mastodon.py`→`mastodon` (16), `hyperion-py`→`hyperion` (14) |

### Top 10 by missed import count

| PyPI Package | Import Name | Missed Files |
|-------------|-------------|-------------|
| `py-unifi-access` | `uiprotect` | 96 |
| `matter-python-client` | `matter_server` | 93 |
| `pyswitchbot` | `switchbot` | 76 |
| `hap-python` | `pyhap` | 55 |
| `pyserial` | `serial` | 38 |
| `pykodi` | `pykodi` (but imported as submodule) | 32 |
| `pyatv` | `pyatv` | 28 |
| `pillow` | `PIL` | 20 |
| `google-genai` | `google.genai` | 19 |
| `mastodon.py` | `mastodon` | 16 |

home-assistant/core is the ultimate stress test for PyPI name mapping — 1,100+ dependencies with every naming convention in the Python ecosystem.

---
*Added by `/diet-fuzz` run 2026-04-11T1310*

**Comment by kotakanbe** [comment]
Resolved by #275 (wheel-based fallback for PyPI package-to-import name resolution). All PyPI name mismatch cases listed in this issue (pyyaml→yaml, pillow→PIL, beautifulsoup4→bs4, etc.) are now automatically resolved by downloading the wheel and extracting import names from top_level.txt. The remaining click-didyoumean case is a call site detection issue tracked in #242, not a name resolution gap.
